colossalai/zero/sharded_model/sharded_model_v2.py

```diff
@@ -161,13 +169,30 @@ def _reduce_scatter_callback(self, param: Parameter, reduced_grad: torch.Tensor)
         if self.gradient_postdivide_factor > 1:
             # Average grad by world_size for consistency with PyTorch DDP.
             reduced_grad.data.div_(self.gradient_postdivide_factor)
-        # Cast grad to param's dtype (typically FP32). Note: we do this
-        # before the cpu offload step so that this entire hook remains
-        # non-blocking. The downside is a bit more D2H transfer in that case.
-        if self.mixed_precision:
-            orig_param_grad_data = reduced_grad.data
-            reduced_grad.data = reduced_grad.data.to(dtype=param.ca_attr.origin_dtype)
-            # Don't let this memory get reused until after the transfer.
-            orig_param_grad_data.record_stream(torch.cuda.current_stream())
-
-        param._sharded_grad.reduce_scatter_callback(reduced_grad)
+
+        # Make sure we store fp32 grad
+        reduced_grad.data = cast_tensor_to_fp32(reduced_grad.data)
```

Comment:
Where does tensor reduced_grad come from?

Reply:
When dp size = 1, `reduced_grad` is locally computed final gradient. When dp size > 1, `reduced_grad` is the gradient shard after `reduce-scatter`.

Reply:
I see